src/services/item/plugins/app/ws/hooks.ts

```diff
@@ -0,0 +1,181 @@
+import { FastifyPluginAsync } from 'fastify';
+
+import { AppDataVisibility, ItemType, PermissionLevel, Websocket } from '@graasp/sdk';
+
+import { buildRepositories } from '../../../../../utils/repositories';
+import { validatePermission } from '../../../../authorization';
+import { WebsocketService } from '../../../../websockets/ws-service';
+import { ItemService } from '../../../service';
+import { AppActionService } from '../appAction/service';
+import { AppDataService } from '../appData/service';
+import { AppSettingService } from '../appSetting/service';
+import {
+  AppActionEvent,
+  AppDataEvent,
+  AppSettingEvent,
+  appActionsTopic,
+  appDataTopic,
+  appSettingsTopic,
+} from './events';
+
+/**
+ * helper to register app data topic
+ */
+function registerAppDataTopic(
+  websockets: WebsocketService,
+  appDataService: AppDataService,
+  itemService: ItemService,
+) {
+  websockets.register(appDataTopic, async (req) => {
+    const { channel: id, member } = req;
+    const repositories = buildRepositories();
+    const item = await itemService.get(member, repositories, id);
+    await validatePermission(repositories, PermissionLevel.Read, member, item);
+    if (item.type !== ItemType.APP) {
+      throw new Websocket.AccessDeniedError('item is not app');
+    }
+  });
+
+  // on post app data, notify apps of new app data
+  appDataService.hooks.setPostHook('post', async (member, repositories, { appData, itemId }) => {
+    if (itemId !== undefined && appData.visibility === AppDataVisibility.Item) {
+      websockets.publish(appDataTopic, itemId, AppDataEvent('post', appData));
+    }
+  });
+
+  appDataService.hooks.setPostHook('patch', async (member, repositories, { appData, itemId }) => {
+    if (itemId !== undefined && appData.visibility === AppDataVisibility.Item) {
+      websockets.publish(appDataTopic, itemId, AppDataEvent('patch', appData));
+    }
+  });
+
+  appDataService.hooks.setPostHook('delete', async (member, repositories, { appData, itemId }) => {
+    if (itemId !== undefined && appData.visibility === AppDataVisibility.Item) {
+      websockets.publish(appDataTopic, itemId, AppDataEvent('delete', appData));
+    }
+  });
+}
+
+/**
+ * helper to register app action topic
```

Comment:
I think it would make sense to divide this file in 3 parts.

Reply:
You could but I think it's fine